src/std/misc/string.ss

```diff
@@ -254,3 +256,50 @@
    ((? (and number? inexact?) v) "~f")
    ((? (or list? hash-table? vector? ##values? obj-pr?) v) "~r")
    (else "~a")))
+
+;; Like CL SUBSTITUTE but specialized for strings and chars
+(def (string-substitute-char-if
+      newchar predicate string
+      start: (start #f)
+      end: (end #f)
+      from-end: (from-end? #f)
+      count: (count #f)
+      in-place: (in-place? #f))
+  (unless start (set! start 0))
+  (unless end (set! end (string-length string)))
+  (def s (if in-place? string (string-copy string)))
+  (let/cc return
+    (cond
+     ((equal? count 0) (return))
+     (count
+      (for (i (if from-end? (in-range (1- end) (1- start) -1) (in-range start end)))
+        (when (predicate (string-ref s i))
+          (string-set! s i newchar)
+          (decrement! count)
+          (when (zero? count) (return)))))
+     (else
+      (for (i (in-range start end))
+        (when (predicate (string-ref s i))
+          (string-set! s i newchar))))))
+  s)
+
+(def (string-substitute-char
+      newchar oldchar string
+      test: (test #f)
+      test-not: (test-not #f)
+      key: (key #f)
+      start: (start #f)
+      end: (end #f)
+      from-end: (from-end? #f)
+      count: (count #f)
+      in-place: (in-place? #f))
+  (def key_ (or key identity))
```

Comment:
just do `(let (key (or key identity)) ...)` for this, no need to pollute the lexical scope, plus it is bug prone.

Reply:
A passing thought: maybe we should have a body sliced let which does this without letrec semantics, so that you can rename things easily without rightwards drift.

What do you think about supporting this form in some future version of Gerbil:
```
(def (string-substitute-char ...)
  (let key (or key identity))
  (def predicace ...)
  ...)
```

Reply:
maybe it is confusing with named lets however, perhaps we should call it something else.

Reply:
I propose this syntax:
```
(def (string-substitute-char ...)
  (let key =: key) ; notice the keyword to make the discriminant
  ...)
```

Reply:
follow up in #845